### DISCUSSION TITLE
Using @middy/core 3.x with aws-nodejs-typescript

### DISCUSSION THREAD

**Discussion by reikje** [serverless/serverless#11145] [General]
Hello, I tried bumping dependencies on a new project, that was recently created using the `aws-nodejs-typescript` template.

```
ncu -u
[====================] 24/24 100%

 @aws-sdk/client-s3              ^3.95.0  →  ^3.105.0     
 @middy/core                      ^2.5.3  →    ^3.0.4     
 @middy/http-error-handler        ^3.0.3  →    ^3.0.4     
 @middy/http-json-body-parser     ^2.5.3  →    ^3.0.4     
 @middy/validator                 ^3.0.3  →    ^3.0.4     
 @serverless/typescript           ^3.0.0  →   ^3.18.0     
 @types/aws-lambda              ^8.10.71  →  ^8.10.99     
 @types/jest                     ^27.5.1  →   ^28.1.1     
 @types/node                   ^14.18.18  →  ^17.0.41     
 esbuild                        ^0.14.11  →  ^0.14.43     
 eslint                          ^8.16.0  →   ^8.17.0     
 jest                            ^28.1.0  →   ^28.1.1     
 json-schema-to-ts                ^1.5.0  →    ^2.5.3     
 serverless                       ^3.0.0  →   ^3.19.0     
 serverless-esbuild              ^1.23.3  →   ^1.30.0     
 ts-jest                         ^28.0.2  →   ^28.0.4     
 ts-node                         ^10.4.0  →   ^10.8.1     
 tsconfig-paths                   ^3.9.0  →    ^4.0.0     
 typescript                       ^4.1.3  →    ^4.7.3
```

I am hitting a problem with `middy`. My `schema.ts` looks like this:

```
export const inputSchema = {
  type: 'object',
  properties: {
    body: {
      type: 'object',
      properties: {
        input: { type: 'string' },
      },
      required: ['input'],
    },
  },
};
```

The `handler.ts` has this:

```
const previewHandler: ValidatedEventAPIGatewayProxyEvent<typeof inputSchema> = async (event) => {
  const { input } = event.body;  // now gives a TS error
```

This worked fine with the dependency versions in the `aws-nodejs-typescript` template, but after trying to bump the versions I get the following transpiler problem:

```
src/functions/preview/handler.ts:35:13 - error TS2339: Property 'input' does not exist on type '{}'.

    35     const { input } = event.body;
```

Looking for help on what needs to be changed to make this work again.

**Comment by nponeccop** [comment]
This is unrelated to Serverless but a `middy`/`Typescript` problem. `typeof  inputSchema` is not the right type to use. `inputSchema` is a JSON schema object to use in the validator middleware. 

If you want to use both schemas and typed handlers, you should generate the types from the schema using one of the available tools. See https://dev.to/simonireilly/typescript-with-the-openapi-specification-b5e for an introductory discussion.

Oh, also `^14.18.18  →  ^17.0.41` is wrong. The major version of `@types/node` must match the runtime you are targeting. So, bump it to the last `14.x` version first. And then, when you change the runtime to 16, you can bump `@types/node` to 16. There's no runtime 17.x on Amazon.

As a diagnostic measure, I also suggest not to bump the typescript. It's possible that the old code worked because the older compiler accepted wrong code.
